library/summary.ml

```diff
@@ -221,4 +221,69 @@ let ref ?(stage=Stage.Interp) ?(local=false) ~name x =
     in
     r
 
+(** Observables *)
+module type OBSERVABLE =
+sig
+  type token
+  type value
+
+  val register : name:string -> ?override:bool -> value -> token
+
+  val activate : token -> unit
+  val deactivate : token -> unit
+end
+
+module type OBSERVABLE_USER =
+sig
+  include OBSERVABLE
+
+  val is_active : token -> bool
+
+  val all_active : unit -> (string * value) list
+end
+
+module Make
+    (Obs : sig
+       type value
+       val name : string
+     end) : OBSERVABLE_USER with type value = Obs.value =
+struct
+  type token = string
+  type value = Obs.value
+
+  let observers = Stdlib.ref CString.Map.empty
+  let active_observers = ref ~name:Obs.name []
+
+  let register ~name ?(override=false) value =
+    if not override && CString.Map.mem name !observers then
+      CErrors.anomaly Pp.(str Obs.name ++ str " observer " ++
+                          str name ++ str " already exists")
+    else
+      observers := CString.Map.add name value !observers ;
+    name
+
+  let remove name = Util.List.remove String.equal name !active_observers
+
+  let activate name =
+    assert (CString.Map.mem name !observers);
+    active_observers := name :: remove name;
+    ()
+
+  let deactivate name =
+    active_observers := remove name;
+    ()
+
+  let is_active : token -> bool =
+    fun tkn -> List.mem tkn !active_observers
+
+  let all_active () : (string * value) list =
+    (* NOTE: this is not very efficient. *)
+    let all = CString.Map.bindings !observers in
+    let f (k,v) =
+      if is_active k then Some (k, v) else None
+    in
+    List.filter_map f all
```

Comment:
```suggestion
  let all_active () : (token * value) list =
    List.map (fun k -> k, CString.Map.get k !observers) !active_observers
```
may not be more efficient (except in rare cases when there are many observers but not many active ones) but is easier to read IMO